File: packages/wire/transfer_set.go
package wire

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

// CompletionDigest preserves the original single-file token and commits an ordered multi-file
// set by hashing its newline-delimited canonical per-file digests.
func CompletionDigest(files []FileEntry) string {
	if len(files) == 1 {
		return files[0].FileDigest
	}
	digests := make([]string, len(files))
	for i, file := range files {
		digests[i] = file.FileDigest
	}
	sum := sha256.Sum256([]byte(strings.Join(digests, "\n")))
	return hex.EncodeToString(sum[:])
}

File: packages/protocol/src/transfer-set.ts
import { bytesToHex, utf8 } from './bytes.js';
import type { FileEntry } from './transfer.js';
import { sha256 } from './webcrypto.js';

/**
 * Completion token for an ordered file set. A single file retains the original wire value;
 * multiple files hash the newline-delimited canonical per-file digests.
 */
export async function completionDigest(files: readonly FileEntry[]): Promise<string> {
  if (files.length === 1) return files[0]!.fileDigest;
  return bytesToHex(await sha256(utf8(files.map((file) => file.fileDigest).join('\n'))));
}
